### ISSUE TITLE
PyOpenAL works through IDLE 3.7.9, but doesn't work from Visual Studio 2019 or Double-Clicking python file.

### ISSUE THREAD

**Issue by twisted-nematic57** [Zuzu-Typ/PyOpenAL#14]
*.wav file plays when program is run through through IDLE, but if I double-click the python file or use Visual Studio, the console/IDLE prints:

`AL lib: (EE) SetChannelMap: Failed to match front-center channel (2) in channel map`
`AL lib: (EE) alc_cleanup: 1 device not closed`

I am using the following modules in my project:
tkinter
turtle
Pillow 8.0.1
time
PyOpenAL 0.7.11a1
random
sys
PyOpenGL 3.1.5

I am using Python version 3.7.9, 64-bit.

If it helps, my system specs are:
Core i5 10th gen, 64-bit
8GB RAM
Intel UHD graphics + Nvidia GeForce GTX 1650 Ti (Notebook)
512GB SSD

**Comment by Zuzu-Typ** [comment]
Hi there (: 

I haven't been working on this library for almost 2 years, so I got a little rusty.. 😬

> AL lib: (EE) SetChannelMap: Failed to match front-center channel (2) in channel map

I know I've seen this error message before, but unfortunately I don't remember what causes it.
Does anything break or is it just printing out the error message?

> AL lib: (EE) alc_cleanup: 1 device not closed

This one _should_ be fixed by executing `oalQuit()` before exiting your program.

Sorry that I couldn't provide you with more information.

Cheers,
--Zuzu_Typ--

**Comment by twisted-nematic57** [comment]
I have done some experimenting, and it seems that:
1. When LOADING (not PLAYING) a sound file, the errors are printed. Simply importing the PyOpenAL library, however, doesn't output any errors.
2. If I try to import a .flac, .ogg or .opus file with PyOgg installed, then the music plays for a bit and turns into deafening white noise, even through IDLE 3.7.9.
3. The program works just fine (as intended) in IDLE. I need to find out what IDLE does to program before executing it to prevent this error.

"AL lib: (EE) SetChannelMap: Failed to match front-center channel (2) in channel map" is **not** printed when run through IDLE.
"AL lib: (EE) alc_cleanup: 1 device not closed" is **not** printed when run through IDLE.

I have a question:
Since this a 3D sound library, is 3D sound always used?
Does Mono channel audio only work with 3D space, or will 7.1 Surround work in 3D space? I think that the number of channels that the audio file has matters to the error.

Also, please try referencing on the Web for this error, because whenever I search for this error on Google or Bing, all I get is chat rooms filled with complaints/bug reports on a software called "Blender". Also Blender is made in C++ so that wouldn't be much help to me. I need help from a pro.

Cheers,
superhuman2020

**Comment by Zuzu-Typ** [comment]
Ah, yes I think you're right. OpenAL always uses 3D sound, unless you're using stereo sound sources, in which case they will play back without any attenuation. So the sound tracks you use must _always_ be in mono. Otherwise the software doesn't know which channel to use for playback.
That's probably what's causing the error (and possibly also why there's the noise).

Output-wise I think OpenAL only supports stereo (or mono, but that would kind of defeat the point). 
If I read that correctly OpenAL-Soft does technically support surround, but I don't know how difficult that would be to set up.

Well, this library also depends on a build of OpenAL (or OpenAL-Soft), which itself is written in C. PyOpenAL only provides bindings to it as well as some convenience functions. So most of the actual work is done in the C library. And I don't know an awful lot about that one. It's also not super well documented, so it's a little tough to get into.

So I think that using files with just a single channel (i.e. mono) should fix the issue.

Cheers
--Zuzu_Typ--

**Comment by twisted-nematic57** [comment]
Haha, I should have tried that in the first place!
Thanks for the help. Will close issue if it works.

Thanks,
superhuman2020

**Comment by twisted-nematic57** [comment]
Nope, it didn't work.

Mono channels used.

I can give full details about my audio. They are in the image.
![image](https://user-images.githubusercontent.com/70416002/109432522-73e4fa80-79c0-11eb-9817-c26821a86e6e.png)

**Comment by twisted-nematic57** [comment]
Ok, I'll temporarily switch to a 1D sound library instead. Take your time to solve this problem if you want to.

I'm not angry, it's just that this game really needs to succeed compiling, because I want to distribute it to my friends.

Good luck,
superhuman2020

**Comment by Zuzu-Typ** [comment]
Sorry, but since I was unable to reproduce the issue on my end and lack the in-depth knowledge of OpenAL required to help you find the source of the issue, there's not much I can do.

I'll close the issue then if you decided not to investigate any further.

I think [simpleaudio](https://simpleaudio.readthedocs.io/en/latest/index.html) is a good choice as an audio playback package.
PyAudio is also good, but unfortunately is no longer maintained.
Or you could use pygame's mixer, should you already use pygame anyway.

Good luck with your game.

Cheers
--Zuzu_Typ--

**Comment by twisted-nematic57** [comment]
Wait, so playing the .wav file actually plays through:
1. IDLE 3.7.9
2. Visual Studio 2019
3. Compiled python file
4. cx_freeze executable

But if I'm buggin' you too much just leave it. 1D sound is fine. I only wanted 3D sound cuz my game is 3D (PyOpenGL).

Thanks,
superhuman2020

**Comment by twisted-nematic57** [comment]
Lol, all I needed to do was include PyOpenAL as an "include" in my Python compiler. That way, the first error message still prints BUT the sound plays fine and it works.

Just if you want to check my game out:
https://github.com/superhuman2020/mcpy

**Comment by Zuzu-Typ** [comment]
Oh, interesting. 
I didn't try using any compiler besides PyInstaller yet, which does a pretty good job at determining all requirements.

Well I'm glad it works for you now.

I still haven't found a way to reproduce the error, though I know I've seen it before.

Haha, yeah. I also tried making a Minecraft clone once, though I stopped working on it at some point.
This is what it looked like:
![The BlockKing 2021-03-02 20-03-11-071](https://user-images.githubusercontent.com/20467351/109701069-e6a5cf80-7b92-11eb-9c22-85f39bb7cdad.jpg)
